compiler/rustc_mir_transform/src/check_finalizers.rs

```diff
@@ -189,44 +212,60 @@ impl<'a, 'tcx> Visitor<'tcx> for ProjectionChecker<'a, 'tcx> {
-                            err.span_label(
-                                span,
-                                "caused by the expression in `fn drop(&mut)` here because",
-                            );
-                            err.span_label(span, "it uses another `Gc` type.");
-                            err.help("`Gc` finalizers are unordered, so this field may have already been dropped. It is not safe to dereference.");
-                        } else {
-                            err.span_label(
-                                self.cx.arg,
-                                "has a drop method which cannot be safely finalized.",
-                            );
-                            err.span_label(
-                                span,
-                                "caused by the expression in `fn drop(&mut)` here because",
-                            );
-                            err.span_label(
-                                span,
-                                "it uses a type which is not safe to use in a finalizer.",
-                            );
-                            err.help("`Gc` runs finalizers on a separate thread, so drop methods\nmust only use values whose types implement `Send + Sync` or `FinalizerSafe`.");
-                        }
-                        err.emit();
+                        self.emit_err(ty, span);
                     }
                 }
                 _ => (),
             }
         }
         self.super_projection(place_ref, context, location);
     }
+
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        if let TerminatorKind::Call { ref args, .. } = terminator.kind {
+            for caller_arg in self.body.args_iter() {
+                let recv_ty = self.body.local_decls()[caller_arg].ty;
+                for arg in args.iter() {
+                    let arg_ty = arg.ty(self.body, self.cx.tcx);
+                    if arg_ty == recv_ty {
+                        // Currently, we do not recurse into function calls
+                        // to see whether they access `!FinalizerSafe`
+                        // fields, so we must throw an error in `drop`
+                        // methods which call other functions and pass
+                        // `self` as an argument.
+                        //
+                        // Here, we throw an error if `drop(&mut self)`
+                        // calls a function with an argument that has the
+                        // same type as the drop receiver (e.g. foo(x:
+                        // &Self)). This approximation will always prevent
+                        // unsound `drop` methods, however, it is overly
+                        // conservative and will prevent correct examples
+                        // like below from compiling:
+                        //
+                        // ```
+                        // fn drop(&mut self) {
```

Comment:
We should add this example to the test suite and add a similar comment along the lines of "this program should work but our underapproximation causes it not to". That way, if we ever fiddle with this code, we'll know if we've fixed or broken programs we thought should/shouldn't work!